### PR TITLE
DrivAerML: Faster cosine cycling T_max=10 at 4L/512d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The baseline uses T_max=30. On AirfRANS, shorter cycles (T_max=5) beat longer ones (T_max=10). DrivAerML might also benefit from faster cycling — T_max=10 gives 3x more cosine restarts per epoch, more opportunities to find deeper basins.

## Instructions
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 10 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent eren \
  --wandb-name "eren/drivaerml-4L512d-tmax10" \
  --wandb-group "drivaerml-4L512d-tmax-sweep"
```

Report val_primary/surface_rel_l2_pct, best epoch, total epochs.

## Baseline
Current best: **4.619%** at epoch 256 (4L/512d, T_max=30, 267 epochs)
External target: <3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT eren:

## Results

### DrivAerML: T_max=10 at 4L/512d

| Metric | T_max=10 (this run) | T_max=30 (baseline) | Change |
|--------|---------------------|---------------------|--------|
| **val_primary/surface_rel_l2_pct** | **10.333%** (ep 159) | **4.619%** (ep 256) | +124% (2.24x worse) |
| test_primary/surface_rel_l2_pct | 25.849% (at final ep, diverged — unreliable) | 6.244% | N/A |
| Total epochs | 268 | 267 | Comparable throughput |
| Best epoch | 159 | 256 | Earlier, then catastrophic divergence |

- **External target:** <3.71%
- **W&B run:** [3wfb1zbm](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3wfb1zbm)
- **Peak GPU memory:** ~6.2 GB (GPU 0)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent eren --wandb-name "eren/drivaerml-4L512d-tmax10" --wandb-group "drivaerml-4L512d-tmax-sweep"
```

### What happened

**Hypothesis falsified.** T_max=10 is significantly worse than T_max=30 for DrivAerML.

The run hit 10.33% at epoch 159 — then suffered catastrophic instability and never recovered. Key problems:

1. **Frequent catastrophic divergence events:** At least 4 major instability cascades (epochs 112, 162-165, 181-183, 195-198+). During these, val metric spiked to 50-70% and loss exploded 10-20x. Each cascade took 10-20 epochs to partially recover.

2. **Progressive degradation after divergence:** Pre-first-divergence trough envelope was descending cleanly (18%→17%→16%→14%→11%). After the epoch 162 divergence, the model never returned to its pre-divergence trajectory — recovery plateaued around 13-15% before diverging again.

3. **Much higher oscillation amplitude:** T_max=10 produced 5-10% peak-trough swings in normal operation, plus 20-60% catastrophic spikes. T_max=30 baseline has much smoother convergence reaching 4.619%.

4. **Fewer effective training epochs:** Despite running 268 epochs (comparable to baseline's 267), only the first ~160 epochs were productive. The remaining ~100 epochs were spent in divergence/recovery cycles.

**Why T_max=10 fails on DrivAerML but works on AirfRANS:** DrivAerML has larger meshes (50k surface points), higher-dimensional output, and more complex geometry (full 3D vehicles vs 2D airfoils). The aggressive LR cycling at T_max=10 causes gradient magnitude spikes that the 4L/512d model cannot absorb — the larger model capacity amplifies rather than dampens the instability. AirfRANS benefits from rapid cycling because its simpler geometry creates a smoother loss landscape where aggressive exploration finds deeper basins without destabilizing.

### Suggested follow-ups

- **T_max=15 or T_max=20 at 4L/512d:** If faster cycling than T_max=30 is desired, a more moderate reduction may avoid the instability while still increasing restart frequency.
- **T_max=10 + grad-clip:** Gradient clipping (gc=1.0 or 1.5) might prevent the catastrophic spikes that T_max=10 causes, combining faster cycling with stability.
- **T_max=10 at smaller model (3L/192d):** The instability may be model-size-dependent — smaller models may tolerate faster cycling better.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2804 because ADVISOR: Closing — your T_max=10 result of 10.333% diverged, far from the 4.619% baseline. Shorter cosine cycles destabilize DrivAerML training. T_max=30 remains the right setting for this dataset. Thanks, eren..
